File: mpc-back/src/models/WaitingListEntry.ts
import moongose, { Document, Schema } from "mongoose";

interface DayAvailability {
  available: boolean;
  allDay?: boolean;
  startTime?: string;
  endTime?: string;
}

const DayAvailabilitySchema = new Schema(
  {
    available: { type: Boolean, default: false },
    allDay: { type: Boolean, default: false },
    startTime: { type: String, default: "" },
    endTime: { type: String, default: "" },
  },
  { _id: false }
);

interface WeeklyAvailability {
  monday: DayAvailability;
  tuesday: DayAvailability;
  wednesday: DayAvailability;
  thursday: DayAvailability;
  friday: DayAvailability;
  saturday: DayAvailability;

  [key: string]: DayAvailability;
}

export interface IWaitingListEntry extends Document {
  firstName: string;
  lastName: string;
  email: string;
  dateApplied: Date;
  approvalStatus: "pending" | "approved" | "rejected";
  approvedDate?: Date;
  age: number;
  weeklyAvailability: WeeklyAvailability;
}

const WaitingListEntrySchema = new Schema<IWaitingListEntry>({
  firstName: { type: String, required: true },
  lastName: { type: String, required: true },
  email: { type: String, required: true, unique: false },
  dateApplied: { type: Date, default: Date.now },
  approvalStatus: {
    type: String,
    enum: ["pending", "approved", "rejected"],
    default: "pending",
  },
  approvedDate: { type: Date },
  age: { type: Number, required: true },
  weeklyAvailability: {
    monday: {
      type: DayAvailabilitySchema,
      required: true,
      default: { available: false },
    },
    tuesday: {
      type: DayAvailabilitySchema,
      required: true,
      default: { available: false },
    },
    wednesday: {
      type: DayAvailabilitySchema,
      required: true,
      default: { available: false },
    },
    thursday: {
      type: DayAvailabilitySchema,
      required: true,
      default: { available: false },
    },
    friday: {
      type: DayAvailabilitySchema,
      required: true,
      default: { available: false },
    },
    saturday: {
      type: DayAvailabilitySchema,
      required: true,
      default: { available: false },
    },
    sunday: {
      type: DayAvailabilitySchema,
      required: true,
      default: { available: false },
    },
  },
});

const WaitingListEntry = moongose.model<IWaitingListEntry>(
  "WaitingListEntry",
  WaitingListEntrySchema
);

export { WeeklyAvailability, DayAvailability, WaitingListEntry };

File: mpc_admin_app/lib/app/models/WaitingListEntry.dart
class WaitingListEntry {
  String id;
  String firstName;
  String lastName;
  String email;
  DateTime dateApplied;
  String approvalStatus;
  int age;
  WeeklyAvailability weeklyAvailability;

  WaitingListEntry({
    required this.id,
    required this.firstName,
    required this.lastName,
    required this.email,
    required this.dateApplied,
    required this.approvalStatus,
    required this.age,
    required this.weeklyAvailability,
  });

  factory WaitingListEntry.fromJson(Map<String, dynamic> json) {
    return WaitingListEntry(
      id: json['_id'],
      firstName: json['firstName'],
      lastName: json['lastName'],
      email: json['email'],
      dateApplied: DateTime.parse(json['dateApplied']),
      approvalStatus: json['approvalStatus'],
      age: json['age'],
      weeklyAvailability: WeeklyAvailability.fromJson(
        json['weeklyAvailability'],
      ),
    );
  }
}

class WeeklyAvailability {
  List<DayAvailability> days;

  WeeklyAvailability({required this.days});

  factory WeeklyAvailability.fromJson(Map<String, dynamic> json) {
    List<DayAvailability> days = [];
    if (json['monday']['available'] == true) {
      days.add(
        DayAvailability.fromJson(json['monday']).copyWith(day: "Monday"),
      );
    }
    if (json['tuesday']['available'] == true) {
      days.add(
        DayAvailability.fromJson(json['tuesday']).copyWith(day: "Tuesday"),
      );
    }

    if (json['wednesday']['available'] == true) {
      days.add(
        DayAvailability.fromJson(json['wednesday']).copyWith(day: "Wednesday"),
      );
    }
    if (json['thursday']['available'] == true) {
      days.add(
        DayAvailability.fromJson(json['thursday']).copyWith(day: "Thursday"),
      );
    }
    if (json['friday']['available'] == true) {
      days.add(
        DayAvailability.fromJson(json['friday']).copyWith(day: "Friday"),
      );
    }

    if (json['saturday']['available'] == true) {
      days.add(
        DayAvailability.fromJson(json['saturday']).copyWith(day: "Saturday"),
      );
    }

    return WeeklyAvailability(days: days);
  }
}

class DayAvailability {
  bool available;
  bool allDay;
  String day;
  String startTime;
  String endTime;

  DayAvailability({
    this.available = false,
    this.allDay = false,
    this.day = "",
    this.startTime = "",
    this.endTime = "",
  });

  factory DayAvailability.fromJson(Map<String, dynamic> json) {
    return DayAvailability(
      available: json['available'],
      allDay: json['allDay'],

      startTime: json['startTime'],
      endTime: json['endTime'],
    );
  }

  // copy with method
  DayAvailability copyWith({
    bool? available,
    bool? allDay,
    String? day,
    String? startTime,
    String? endTime,
  }) {
    return DayAvailability(
      available: available ?? this.available,
      allDay: allDay ?? this.allDay,
      day: day ?? this.day,
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
    );
  }
}
